src/backend/distributed/operations/repair_shards.c

```diff
@@ -158,6 +173,120 @@ master_copy_shard_placement(PG_FUNCTION_ARGS)
+																	  TOTAL_RELATION_SIZE,
+																	  optimizePartitionCalculations);
+
+	MultiConnection *connection = GetNodeConnection(connectionFlag, workerNodeName,
+													workerNodePort);
+	PGresult *result = NULL;
+	int queryResult = ExecuteOptionalRemoteCommand(connection, tableSizeQuery->data,
+												   &result);
+
+	if (queryResult != RESPONSE_OKAY)
+	{
+		ereport(ERROR, (errcode(ERRCODE_CONNECTION_FAILURE),
+						errmsg("cannot get the size because of a connection error")));
+	}
+
+	List *sizeList = ReadFirstColumnAsText(result);
+	if (list_length(sizeList) != 1)
+	{
+		ereport(ERROR, (errmsg(
+							"received wrong number of rows from worker, expected 1 received %d",
+							list_length(sizeList))));
+	}
+
+	StringInfo tableSizeStringInfo = (StringInfo) linitial(sizeList);
+	char *tableSizeString = tableSizeStringInfo->data;
+	uint64 tableSize = SafeStringToUint64(tableSizeString);
+
+	bool raiseErrors = true;
+	PQclear(result);
+	ClearResults(connection, raiseErrors);
+
+	return tableSize;
+}
+
+
+/*
+ * CheckSpaceConstraints checks there is enough space to place the colocation
+ * on the node that the connection is connected to.
+ */
+static
+DeferredErrorMessage *
+CheckSpaceConstraints(MultiConnection *connection, int64 colocationSizeInBytes)
+{
+	uint64 diskAvailableInBytes = 0;
+	uint64 diskSizeInBytes = 0;
+	bool success =
+		GetNodeDiskSpaceStatsForConnection(connection, &diskAvailableInBytes,
+										   &diskSizeInBytes);
+	if (!success)
+	{
+		ereport(ERROR, (errmsg("Could not fetch disk stats for a node")));
+	}
+
+	if (ForceDiskAvailableInBytes != NOT_SET)
+	{
+		diskAvailableInBytes = ForceDiskAvailableInBytes;
+	}
+
+	if (ForceDiskSizeInBytes != NOT_SET)
+	{
+		diskSizeInBytes = ForceDiskSizeInBytes;
+	}
+
+	StringInfo errorDetail = makeStringInfo();
+	if (diskAvailableInBytes < colocationSizeInBytes)
+	{
+		bool isAvailableSpaceForced = (ForceDiskAvailableInBytes != NOT_SET);
+		if (isAvailableSpaceForced)
+		{
+			appendStringInfo(errorDetail,
+							 "actual space %ld bytes (forced), required space %ld bytes",
+							 diskAvailableInBytes, colocationSizeInBytes);
+		}
+		else
+		{
+			appendStringInfo(errorDetail,
+							 "actual space %ld bytes, required space %ld bytes",
+							 diskAvailableInBytes, colocationSizeInBytes);
+		}
+
+		return DeferredError(ERRCODE_INSUFFICIENT_RESOURCES,
+							 "not enough space to move shard",
+							 errorDetail->data, NULL);
+	}
+	int64 newDiskAvailableInBytes = diskAvailableInBytes - colocationSizeInBytes;
+	int64 desiredNewDiskAvailableInBytes = diskSizeInBytes *
+										   (DesiredPercentFreeAfterMove / 100);
+	if (newDiskAvailableInBytes < desiredNewDiskAvailableInBytes)
+	{
+		appendStringInfo(errorDetail, "actual available space after move %ld bytes, "
+									  "desired available space after move %ld bytes",
+						 newDiskAvailableInBytes, desiredNewDiskAvailableInBytes);
+		return DeferredError(ERRCODE_INSUFFICIENT_RESOURCES,
```

Comment:
this error message feels like we moved the shards and we don't have enough space. Can you make it clear that this has not happened, but would happen if we try to move.

And, the hint also can be more descriptive.

Reply:
Here I am also thinking of adding how much the estimated colocation size is so that users can have an idea of how much space is needed